### PR TITLE
rustc: Add a new `-Z force-unstable-if-unmarked` flag

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit adds a new `-Z` flag to the compiler for use when bootstrapping the
compiler itself. We want to be able to use crates.io crates, but we also want
the usage of such crates to be as ergonomic as possible! To that end compiler
crates are a little tricky in that the crates.io crates are not annotated as
unstable, nor do they expect to pull in unstable dependencies.

To cover all these situations it's intended that the compiler will forever now
bootstrap with `-Z force-unstable-if-unmarked`. This flags serves a dual purpose
of forcing crates.io crates to themselves be unstable while also allowing them
to use other "unstable" crates.io crates. This should mean that adding a
dependency to compiler no longer requires upstream modification with
unstable/staged_api attributes for inclusion!

### CODE REVIEW

**Comment by alexcrichton** [comment]
cc @eddyb

File: src/bootstrap/bin/rustc.rs [hunk 1]
```diff
@@ -194,6 +194,8 @@ fn main() {
                 // do that we pass a weird flag to the compiler to get it to do
                 // so. Note that this is definitely a hack, and we should likely
                 // flesh out rpath support more fully in the future.
+                //
+                // TODO: remove condition after next stage0
```
Comment (Mark-Simulacrum) [inline src/bootstrap/bin/rustc.rs]: This is unrelated to this PR's primary changes, right? Just want to make sure...

Comment (alexcrichton) [inline src/bootstrap/bin/rustc.rs]: Indeed! I just remembered to touch it up in this PR

File: src/librustc/middle/stability.rs [hunk 2]
```diff
@@ -390,20 +390,36 @@ impl<'a, 'tcx> Index<'tcx> {
             parent_depr: None,
             in_trait_impl: false,
         };
+
+        // If the `-Z force-unstable-if-unmarked` flag is passed then we provide
+        // a parent stability annotation which indicates that this is private
+        // with the `rustc_private` feature. This is intended for use when
+        // compiling librustc crates themselves so we can leverage crates.io
+        // while maintaining the invariant that all sysroort crates are unstable
+        // by default an are unable to be used.
```
Comment (pitdicker) [inline src/librustc/middle/stability.rs]: *sysroot, *and

File: src/librustc/session/config.rs [hunk 3]
```diff
@@ -1027,6 +1027,8 @@ options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
         "add a source pattern to the file path remapping config"),
     remap_path_prefix_to: Vec<String> = (vec![], parse_string_push, [TRACKED],
         "add a mapping target to the file path remapping config"),
+    force_unstable_if_unmarked: bool = (false, parse_bool, [TRACKED],
+        "force all crates to be `rustc_private` unstable"),
```
Comment (eddyb) [inline src/librustc/session/config.rs]: Could the `rustc_private be provided from rustbuild instead if being hardcoded?

File: src/librustc/middle/stability.rs [hunk 4]
```diff
@@ -496,8 +512,10 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
             }
         }
 
-        let is_staged_api = *self.stability.borrow_mut().staged_api.entry(def_id.krate)
-            .or_insert_with(|| self.sess.cstore.is_staged_api(def_id.krate));
+        let is_staged_api = self.lookup_stability(DefId {
+            index: CRATE_DEF_INDEX,
+            ..def_id
+        }).is_some();
```
Comment (eddyb) [inline src/librustc/middle/stability.rs]: Aah this is what I didn't consider refactoring - it's much nicer!

File: src/librustc/session/config.rs (hunk 3 above)
Comment (alexcrichton) [inline src/librustc/session/config.rs]: We could, yeah, but I'm not sure it'd buy much? We otherwise would still hardcode the issue number as well as the description of the unstable feature (the "note"), and passing all that info as CLI flags may not also be worth it :(.

Do you think we'll want different unstable features on differnet crates though?

Comment (eddyb) [inline src/librustc/session/config.rs]: It's more that the flag name doesn't seem specific to `rustc_private`. I don't care too much anyway.

Comment (alexcrichton) [inline src/librustc/session/config.rs]: Heh we could also call it `-Z tag-this-crate-with-rustc-private-unstable-feature-if-otherwise-unmarked`

**Comment by Mark-Simulacrum** [comment]
Legitimate tidy errors.

```
[00:01:56] Building stage0 tool tidy (x86_64-unknown-linux-gnu)
[00:01:57]    Compiling tidy v0.1.0 (file:///checkout/src/tools/tidy)
[00:02:04]     Finished release [optimized] target(s) in 6.87 secs
[00:02:04] tidy check (x86_64-unknown-linux-gnu)
[00:02:04] tidy error: /checkout/src/bootstrap/bin/rustc.rs:198: TODO is deprecated; use FIXME
[00:02:04] tidy error: /checkout/src/bootstrap/bin/rustc.rs:228: TODO is deprecated; use FIXME
[00:02:04] tidy error: /checkout/src/bootstrap/lib.rs:529: TODO is deprecated; use FIXME
[00:02:05] some tidy checks failed
```

**Comment by alexcrichton** [comment]
Fixed!

**Comment by alexcrichton** [comment]
I think @pnkfelix may be on leave, @eddyb mind reviewing this?

**Comment by Mark-Simulacrum** [comment]
@bors p=1

Giving higher priority since I know this blocks at least a few other PRs.
